hare3/remote_test.go

```diff
@@ -0,0 +1,233 @@
+package hare3
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+	"go.uber.org/zap/zaptest"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/signing"
+)
+
+func TestRemoteHare(t *testing.T) {
+	cfg := Config{
+		Committee:       10,
+		IterationsLimit: 5,
+		PreroundDelay:   time.Millisecond,
+		RoundDuration:   time.Millisecond * 10,
+		ProtocolName:    "rh3",
+	}
+	ctrl := gomock.NewController(t)
+	svc := NewMockNodeService(ctrl)
+	certifier := NewMockcertifier(ctrl)
+	oracle := NewMockoracle(ctrl)
+	nodeClock := testNodeClock{
+		genesis:       time.Now(),
+		layerDuration: time.Minute,
+	}
+
+	hare := NewRemoteHare(cfg, &nodeClock, svc, nil, oracle, certifier, zaptest.NewLogger(t))
+
+	s := &session{
+		lid:     nodeClock.CurrentLayer(),
+		beacon:  types.Beacon{1, 2, 3, 4},
+		signers: make([]*signing.EdSigner, 1),
+		vrfs:    make([]*types.HareEligibility, 1),
+		proto:   newProtocol(10, hare.log.Named("proto")),
+	}
+
+	signer, err := signing.NewEdSigner()
+	require.NoError(t, err)
+	s.signers[0] = signer
+
+	oracle.EXPECT().
+		CalcEligibility(gomock.Any(), s.lid, gomock.Any(), gomock.Any(), signer.NodeID(), gomock.Any()).
+		Return(10, nil).
+		AnyTimes()
+
+	svc.EXPECT().HareRoundTemplate(gomock.Any(), s.lid, IterRound{
+		Iter:  0,
+		Round: preround,
+	}).Return(&Body{}, nil)
+	svc.EXPECT().Publish(gomock.Any(), cfg.ProtocolName, gomock.Any())
+
+	for iter := range uint8(2) {
+		svc.EXPECT().HareRoundTemplate(gomock.Any(), s.lid, IterRound{
+			Iter:  iter,
+			Round: propose,
+		}).Return(&Body{}, nil)
+		svc.EXPECT().HareRoundTemplate(gomock.Any(), s.lid, IterRound{
+			Iter:  iter,
+			Round: commit,
+		}).Return(&Body{}, nil)
+		svc.EXPECT().HareRoundTemplate(gomock.Any(), s.lid, IterRound{
+			Iter:  iter,
+			Round: notify,
+		}).Return(&Body{}, nil)
+		svc.EXPECT().Publish(gomock.Any(), cfg.ProtocolName, gomock.Any()).Times(3)
```

Comment:
This is confusing to me. Isn't publish called 10 times in total because of being eligible for 10 proposals as set by line 46 why the separation into different rounds?

Reply:
The eligibility set to `10` on line 46 has nothing to do with the number of published messages. Hare eligibility says how much voting power given participant has in each round. For example, a single ID vote (on a proposal) in hare might count as 3 votes. For a proposal to be included for the next round, it needs to get at least `commitee_size / 2 + 1` votes. The expectation set on line 46 (`CalcEligibility() -> 10`) merely makes the signer active in every round. The exact number returned doesn't make any difference.

Back to the `EXPECT().Publish()`, hare publishes messages in the following rounds: `preround` (only happens in the first iteration), `propose`, `commit` and `notify`, if the signer is active in given round (in the test the signer is active in every round).